server/devpi_server/views.py

```diff
@@ -1446,6 +1446,104 @@ def user_list(self):
             d[user.name] = user.get()
         apireturn(200, type="list:userconfig", result=d)
 
+    @view_config(
+        route_name="/+snapshot/{user}/{index}", request_method="POST")
+    @view_config(
+        route_name="/+snapshot/{user}/{index}/", request_method="POST")
+    def snapshot(self):
+        request = self.request
+        context = self.context
+        if not request.has_permission("upload"):  # TODO: add 'snapshot' perm ?
+            # if there is no authenticated user, then issue a basic auth challenge
+            if not request.authenticated_userid:
+                response = HTTPUnauthorized()
+                response.headers.update(forget(request))
+                return response
+            abort_submit(request, 403, "no permission to snapshot")
+            assert 0
+        json = getjson(request)
+        target_index = json["target_index"]
+        target_stage = None
+        source_stage, projects = context.stage.list_projects()[0]
+        failures = []
+        notices = []
+        tot_wheels = 0
+        tot_project_versions = 0
+        if context.user.getstage(target_index):
+            apireturn(409, "index %r already exists" % target_index)
+        try:
+            target_stage = context.user.create_stage(target_index)
+            for project_name in projects:
+                print(f"doing {project_name} ..")
```

Comment:
Use logging with regular string formatting and arguments instead of ``%`` operator to avoid creating the string if the log record is discarded by filters.